client/pkg/connectionctx/manager.go

```diff
@@ -0,0 +1,143 @@
+	// Current stream to send the gRPC requests.
+	Stream T
+}
+
+// Manager is used to manage the connection contexts.
+type Manager[T any] struct {
+	sync.RWMutex
+	connectionCtxs map[string]*connectionCtx[T]
+}
+
+// NewManager is used to create a new connection context manager.
+func NewManager[T any]() *Manager[T] {
+	return &Manager[T]{
+		connectionCtxs: make(map[string]*connectionCtx[T], 3),
+	}
+}
+
+// Exist is used to check if the connection context exists by the given URL.
+func (c *Manager[T]) Exist(url string) bool {
+	c.RLock()
+	defer c.RUnlock()
+	_, ok := c.connectionCtxs[url]
+	return ok
+}
+
+// Store is used to store the connection context, `overwrite` is used to force the store operation
+// no matter whether the connection context exists before, which is false by default.
+func (c *Manager[T]) Store(ctx context.Context, url string, stream T, overwrite ...bool) {
+	c.Lock()
+	defer c.Unlock()
+	overwriteFlag := false
+	if len(overwrite) > 0 {
+		overwriteFlag = overwrite[0]
+	}
+	_, ok := c.connectionCtxs[url]
+	if !overwriteFlag && ok {
+		return
+	}
+	c.storeLocked(ctx, url, stream)
+}
+
+func (c *Manager[T]) storeLocked(ctx context.Context, url string, stream T) {
+	c.releaseLocked(url)
+	cctx, cancel := context.WithCancel(ctx)
+	c.connectionCtxs[url] = &connectionCtx[T]{cctx, cancel, url, stream}
+}
+
+// ExclusivelyStore is used to store the connection context exclusively. It will release
+// all other connection contexts. `stream` is optional, if it is not provided, all
+// connection contexts other than the given `url` will be released.
+func (c *Manager[T]) ExclusivelyStore(ctx context.Context, url string, stream ...T) {
+	c.Lock()
+	defer c.Unlock()
+	// Remove all other `connectionCtx`s.
+	c.gcLocked(func(curURL string) bool {
+		return curURL != url
+	})
+	if len(stream) == 0 {
+		return
+	}
+	c.storeLocked(ctx, url, stream[0])
+}
+
+// GC is used to release all connection contexts that match the given condition.
+func (c *Manager[T]) GC(condition func(url string) bool) {
+	c.Lock()
+	defer c.Unlock()
+	c.gcLocked(condition)
+}
+
+func (c *Manager[T]) gcLocked(condition func(url string) bool) {
+	for url := range c.connectionCtxs {
+		if condition(url) {
+			c.releaseLocked(url)
+		}
+	}
+}
+
+// ReleaseAll is used to release all connection contexts.
+func (c *Manager[T]) ReleaseAll() {
+	c.GC(func(string) bool { return true })
+}
+
+// Release is used to delete a connection context from the connection context map and release the resources.
+func (c *Manager[T]) Release(url string) {
+	c.Lock()
+	defer c.Unlock()
+	c.releaseLocked(url)
+}
+
+func (c *Manager[T]) releaseLocked(url string) {
+	cc, ok := c.connectionCtxs[url]
+	if !ok {
+		return
+	}
+	cc.Cancel()
+	delete(c.connectionCtxs, url)
+}
+
+// Choose is used to choose a connection context from the connection context map.
+// It uses the reservoir sampling algorithm to randomly choose a connection context.
+func (c *Manager[T]) Choose() *connectionCtx[T] {
```

Comment:
I prefer `Choose` since the reservoir sampling algorithm is used here to randomly choose a stream when there are two or more, though `GetConnectionCtx()` may fit the case of only one.

Reply:
I think the user do not need to consider the principle